63_CUDAInterop/main.cpp

```diff
@@ -123,89 +155,137 @@ class CUDA2VKApp : public examples::MonoDeviceApplication
-			params.externalHandleTypes = CCUDADevice::EXTERNAL_MEMORY_HANDLE_TYPE;
-			params.tiling = IGPUImage::TILING::LINEAR;
-			importedimg = mem[2]->createAndBindImage(m_device.get(), std::move(params));
-			assert(importedimg);
+			importedImg = cudaMemories[2]->createAndBindImage(m_device.get(), std::move(params));
+			if (!importedImg) logFail("Failed to create an external image");
 		}
 		
-		commandPool = m_device->createCommandPool(queue->getFamilyIndex(), {});
-		bool re = commandPool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1, &cmd, smart_refctd_ptr(m_logger));
-		assert(re);
+		commandPool = m_device->createCommandPool(queue->getFamilyIndex(), IGPUCommandPool::CREATE_FLAGS::RESET_COMMAND_BUFFER_BIT);
+		bool re = commandPool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 2, cmd, smart_refctd_ptr(m_logger));
 
-		auto createStaging = [logicalDevice= m_device]()
+		auto createStaging = [this,logicalDevice= m_device]()
 		{
 			auto buf = logicalDevice->createBuffer({ {.size = size, .usage = asset::IBuffer::EUF_TRANSFER_DST_BIT} });
 			auto req = buf->getMemoryReqs();
 			req.memoryTypeBits &= logicalDevice->getPhysicalDevice()->getDownStreamingMemoryTypeBits();
 			auto allocation = logicalDevice->allocate(req, buf.get());
-			assert(allocation.isValid() && buf->getBoundMemory().memory->isMappable());
 			
-			bool re = allocation.memory->map(IDeviceMemoryAllocation::MemoryRange(0, req.size), IDeviceMemoryAllocation::EMCAF_READ);
-			assert(re && allocation.memory->getMappedPointer());
-			memset(allocation.memory->getMappedPointer(), 0, req.size);
+			void* mapping = allocation.memory->map(IDeviceMemoryAllocation::MemoryRange(0, req.size), IDeviceMemoryAllocation::EMCAF_READ);
+			if (!mapping) 
+				logFail("Failed to map an staging buffer");
+			memset(mapping, 0, req.size);
 			return buf;
 		};
 
-		stagingbuf = createStaging();
-		stagingbuf2 = createStaging();
+		stagingBufs[0] = createStaging();
+		stagingBufs[1] = createStaging();
 	}
 
 	void launchKernel(CUfunction kernel, CUstream stream)
 	{
+
+		// First we record a release ownership transfer to let vulkan know that resources are going to be used in an external API
+		{
+			IGPUCommandBuffer::SBufferMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> bufBarrier = {
+					.barrier = {
+						.ownershipOp = IGPUCommandBuffer::SOwnershipTransferBarrier::OWNERSHIP_OP::RELEASE,
+						.otherQueueFamilyIndex = IQueue::FamilyExternal,
+					},
+					.range = {.buffer = importedBuf, },
+			};
+
+			IGPUCommandBuffer::SImageMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> imgBarrier = {
+				.barrier = {
+					.ownershipOp = IGPUCommandBuffer::SOwnershipTransferBarrier::OWNERSHIP_OP::RELEASE,
+					.otherQueueFamilyIndex = IQueue::FamilyExternal,
+				},
+				.image = importedImg.get(),
+				.subresourceRange = {
+					.aspectMask = IImage::EAF_COLOR_BIT,
+					.levelCount = 1u,
+					.layerCount = 1u,
+				}
```

Comment:
explain why you're not doing the usual UNDEFINED to X layout transition we usually see in samples